Refactor MasterBoardData and RobotModeData publishers/consumers

gavanderhoorn opened the pull request:
This is a refactor of #162, which rearranges responsibilities in a way that (to me) seems more in-line with the design of @Zagitta.

I've only done a build-test, would appreciate if someone (@miguelprada?) could test this against actual hardware.

gavanderhoorn commented:
This actually depended on https://github.com/ros-industrial/universal_robot/pull/416, but only implicitly, as the fields were already there in the message, but just didn't have the correct width (which would have led to issues at runtime, but Travis doesn't detect those).

miguelprada commented:
I currently only have a UR5e, so I had to cherry-pick #247 and your two commits on top of `dniewinski/kinetic-devel` in order to test these changes. When doing this, the masterboard messages are published fine and I can confirm that the bitmasks contain the right digital output information (I had no easy way to set the inputs).

However robot mode data for versions 3.2 and up is not published so I cannot verify that works. What's the rationale for doing that? I saw your comment in the code, but it's not like 3.2 is the first version where fields not present in the `RobotModeDataMsg.msg` were added to the protocol. There are fields in the protocol that exist from PolyScope 1.6 that are not there in the ROS message, e.g. `speedFraction` (renamed later to `speedScaling`?).

I will setup a UR5 next week, but I think that runs 3.5, so I will be in a similar situation with respect to testing the robot mode data publishing.

gavanderhoorn commented:
> However robot mode data for versions 3.2 and up is not published so I cannot verify that works. What's the rationale for doing that? I saw your comment in the code, but it's not like 3.2 is the first version where fields not present in the `RobotModeDataMsg.msg` were added to the protocol. There are fields in the protocol that exist from PolyScope 1.6 that are not there in the ROS message, e.g. `speedFraction` (renamed later to `speedScaling`?).

Reason I did not add version-specific handling there is that the message in `ur_msgs` currently doesn't contain those fields (see [](https://github.com/ros-industrial/universal_robot/blob/f6333f0ccd57ed378b5a110e19ad2b43ffe9e604/ur_msgs/msg/RobotModeDataMsg.msg)). It explicitly states this as well ([here](https://github.com/ros-industrial/universal_robot/blob/f6333f0ccd57ed378b5a110e19ad2b43ffe9e604/ur_msgs/msg/RobotModeDataMsg.msg#L6)):

> Note: this message does not carry all fields from the RobotModeData structure as broadcast by the robot controller, but a subset.

As this PR is only a refactor, I did not want to extend the message def and start parsing version-specific fields.

We can add that later.

miguelprada commented on src/ros/mb_publisher.cpp:
```diff
@@ -148,15 +207,19 @@ bool MBPublisher::consume(MasterBoardData_V3_2& data)
 bool MBPublisher::consume(RobotModeData_V3_2& data)
 {
   publishRobotStatus(data);
+  // TODO: add robot mode v3.2 handler (would require changes to RobotModeData
```
Let me add that by adding `publishRobotModeData(data)` here, I can see the messages being published.

gavanderhoorn replied:
I could add that call to all `consume(..)` overloads. It wouldn't parse any of the additional fields, but it would cause publication of the data that is parsed for all supported versions.

miguelprada replied:
That call is already there for versions other than 3.2. That's why I wondered why you weren't adding it to this version as well.

gavanderhoorn replied:
This was an oversight, not sure what (or whether) I was (still) thinking (t)here.

miguelprada commented:
> Reason I did not add version-specific handling there is that the message in ur_msgs currently doesn't contain those fields (ur_msgs/RobotModeDataMsg.msg).

It's not that there's no version-specific handling, but rather that not even the common parts are published. Maybe it was just an omission? See my [other comment](https://github.com/ros-industrial/ur_modern_driver/pull/281#discussion_r272569974).

gavanderhoorn commented:
@miguelprada: I believe I've addressed your comments (which were completely correct btw).

There's still no version-specific fields for `RobotModeData`, but see https://github.com/ros-industrial/ur_modern_driver/pull/281#issuecomment-480261060.

gavanderhoorn commented:
Thanks @miguelprada for the review :+1:
